src/tools/asm2wasm.cpp

```diff
@@ -131,9 +131,8 @@ int main(int argc, const char *argv[]) {
   }
 
   if (options.runningDefaultOptimizationPasses()) {
-    if (options.passes.size() > 1) {
-      Fatal() << "asm2wasm can only run default optimization passes (-O, -Ox, etc.), and not specific additional passes";
-    }
+    FatalIf(options.passes.size() > 1)
+      << "asm2wasm can only run default optimization passes (-O, -Ox, etc.), and not specific additional passes";
   }
```

Comment:
Heh, this is a cool idea!

I like it, but my main concern is perf. In e.g. the diff here, previously we checked the condition and only if it's true do we create a Fatal object and pass it the string etc. But with the change, we always do that work, I think? Doesn't matter in most places, but we'd need to be careful to avoid it in places that are called a lot, like say the validator. But, if this is only for validating user input, maybe this isn't an issue anyhow as those places are in fact not called a lot.

Reply:
I don't think the work is significantly different. Previously you'd evaluate the condition. Now you'll call the constructor, call operator<<, and call the destructor. But each of those evaluates the condition; I'd be surprised if the optimizer didn't simplify that to a single evaluation for all 3 actions, which would make the result identical.

Reply:
But surely in the case where the condition is false, the old code won't ever call the constructor?

Reply:
Yes, all the work is behind the conditional in the old code. I'm just saying that for the new case after inlining I would expect the compiler to transform the 3 checks (each in one of the methods) to a single one, putting the same work behind the check.

Reply:
I'd expect the these to all get inlined, and then all the `if (condition)`s to get coalesced into one, so these should be identical code-gen wise, though maybe that's relying too much on the optimizer.

The thing I like about this is we still keep the old `Fatal` for places where `FatalIf` is impractical. For semantics: https://github.com/WebAssembly/binaryen/blob/dd95d06c78757775ae52db08064805adaaf13cc3/src/tools/s2wasm.cpp#L247-L251
And I can see making this more explicit for perf-sensitive bits too, to avoid relying on the optimizer.

Reply:
I forgot we could, like, check: https://godbolt.org/g/Kj4Lj4
Looks like nearly-identical codegen to me, thanks inlining.

Reply:
Aside from the checks, though, there is the work of the constructor and the string creation? If we did
```
FatalIf(isBad(func->body)) << "bad function body: " << *func->body;
```
then it would convert the entire function body to the text format even if it isn't needed?